shark_turbine/kernel/wave/scheduling/graph_utils.py

```diff
@@ -136,3 +153,40 @@ def find_cycles_in_graph(
             circuits += circuit(s, B)
             remove_node(graph, s)
     return circuits
+
+
+def all_pairs_longest_paths(
+    graph: fx.Graph,
+    edges: list[Edge],
+) -> dict[tuple[int, int], IndexExpr]:
+    """
+    For each node in the graph, compute the longest path to all other nodes.
+    Uses the Floyd-Warshall algorithm and assumes that the cycles don't
+    have positive weights.
+    """
+    T = index_symbol("$INITIATION_INTERVAL")
+    D: dict[tuple[int, int, int]] = {}
+    for v in graph.nodes:
+        for w in graph.nodes:
+            D[(v, w)] = -math.inf
+    for edge in edges:
+        D[(edge._from, edge._to)] = (
+            edge.weight.delay - edge.weight.iteration_difference * T
+        )
+    for u in graph.nodes:
+        for v in graph.nodes:
+            for w in graph.nodes:
+                D[(v, w)] = sympy.Max(D[(v, w)], D[(v, u)] + D[(u, w)])
```

Comment:
N^3 complexity is unfortunate, but we can live with it for, I suppose.

Reply:
yeah I dont think there's a way around this unfortunately.